docs: soft opt out adr (#1014)
cosmos/interchain-security: docs/docs/adrs/adr-009-soft-opt-out.md

---
sidebar_position: 10
title: Soft Opt-Out
---
## ADR 009: Soft Opt-Out

## Changelog

* 6/13/23: Initial draft of ADR. Feature already implemented and in production.

## Status

Accepted

## Context

> Some small validators may not have the resources needed to validate all consumer chains. Therefore a need exists to allow the bottom _% of validators to opt-out of validating a consumer chain. Meaning downtime infractions for relevant validators are dropped without ever reaching the provider.

Reviewer: ```suggestion
Some small validators may not have the resources needed to validate all consumer chains. Therefore a need exists to allow the bottom `x%` of validators to opt-out of validating a consumer chain. Meaning downtime infractions for these validators are dropped without ever reaching the provider.
```

This document specifies a modification to the ccv protocol which allows the bottom x% of the validator set by power to opt out of validating consumer chains without being jailed or otherwise punished for it. The feature is implemented with entirely consumer-side code.

## Decision

A consumer param exists, known as `SoftOptOutThreshold`, which is a string decimal in the range of [0, 0.2], that determines the portion of validators which are allowed to opt out of validating that specific consumer.

> Every consumer beginblocker, a function is ran which determines the voting power s.t. any validator with greater than or equal to said voting power must validate that specific consumer. Any validator with less than said voting power may opt out of validating that consumer chain.

Reviewer: ```suggestion
In every consumer beginblocker, a function is ran which determines a _threshold voting power_ for the opt-out mechanism. Validators with voting power greater than or equal to this threshold must validate the consumer chain. Validators below the threshold may opt out of validating the consumer chain.
```

Reviewer: ```suggestion
In every consumer beginblocker, a function is ran which determines the so called  _smallest non opt-out voting power_. Validators with voting power greater than or equal to this value must validate the consumer chain, while validators below this value may opt out of validating the consumer chain.
```

> The voting power value described above is referred to in code as "smallest non opt out power". The value is recomputed every beginblocker in `UpdateSmallestNonOptOutPower()`. In a nutshell, the method obtains the total voting power of the consumer, iterates through the full valset (ordered power ascending) keeping track of a power sum, and when `powerSum / totalPower > SoftOptOutThreshold`, the `SmallestNonOptOutPower` is found and persisted.

Reviewer: ```suggestion
The threshold voting power described above is referred to in code as the _**smallest non opt-out power**_. The value is recomputed every beginblocker in `UpdateSmallestNonOptOutPower()`. In a nutshell, the method obtains the total voting power of the consumer, iterates through the full valset (ordered power ascending) keeping track of a power sum, and when `powerSum / totalPower > SoftOptOutThreshold`, the `SmallestNonOptOutPower` is found and persisted.
```

Author: Formatting stuff seems fine, see above on "threshold"

Then, whenever the `Slash()` interface is executed on the consumer, if the voting power of the relevant validator being slashed is less than `SmallestNonOptOutPower` for that block, the slash request is dropped and never sent to the provider.

## Consequences

### Positive

* Small validators can opt out of validating specific consumers without being punished for it.

### Negative

> * Small validators who are not being slashed will still be present in the validator set of the consumer chain. In other words, we are sacrificing liveness, not safety. The economic security of the consumer chain is the same, but consumers may be more likely to halt. If soft opt out threshold is set to 10% for example, and every validator who doesn't have to validate the consumer doesn't validate it, then 23% downtime of the remaining valset could halt the chain. This may manifest in slightly longer downtime periods during upgrades.

Reviewer: ```suggestion
The bottom `x%` is still part of the total voting power of the consumer chain. This means that if the soft opt-out threshold is set to `10%` for example, and every validator in the bottom `10%` opts out from validating the consumer, then a `24%` downtime of the remaining voting power would halt the chain. This may be especially problematic during consumer upgrades.
```

### Neutral

> * In nominal scenarios, consumers with soft opt out enabled will be computing slash packets for small vals, which may be dropped. This is wasted computation, but necessary to keep implementation simple.

Reviewer: As we've seen in the case of Neutron, this is quite bad. I'd move to Negative with a pointer that the entire downtime logic in the slashing module is executed.
